# Import Packages

In [1]:
import pandas as pd
import numpy as np
import random

# repeated printouts
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

# Load the raw data

We are only loading the excel sheets we will use for our research purposes. The sheets we need are `Demographics`, `Presenting Problems`, `Emergency Crisis Services`.

In [2]:
# Read selected excel sheets into pandas dataframes and save in a dictionary
file_path = '../data/Dartmouth Data Set.xlsx'
sheet_names = ['Demographics', 'Presenting Problems']

# create an empty dictionary to store the sheet data
sheet_data = {}

# loop through each sheet name and read the corresponding sheet data into a pandas dataframe
for sheet_name in sheet_names:
    sheet_data[sheet_name] = pd.read_excel(file_path, sheet_name)

# Save the data in variables and inspect them

In [3]:
# Retrieve the dataframes from the dictionary and save them in variables
demographics = sheet_data['Demographics']
presenting_problems = sheet_data['Presenting Problems']

In [4]:
# Explore the demographics dataframe
demographics.head()
demographics.columns
demographics.info()

Local ID                             Region        County  \
0  8085297     California : CA START East Bay  Contra Costa   
1   440987                New York : Region 3      Saratoga   
2  8008815     California : CA START East Bay       Alameda   
3   220396                New York : Region 3        Albany   
4  6570649  California : CA START San Andreas    Santa Cruz   

     Start Coordinator Date Enrolled in START  \
0        Mary Akhuetie             2020-12-30   
1      Natalia Wolosen             2020-12-29   
2  Sharon Cyrus-Savary             2020-12-29   
3      Matthew Wolcott             2020-12-29   
4        Ilian Navarro             2020-12-29   

                                     Status Status Date  \
0                                    Active  2020-12-30   
1             Inactive (stable functioning)  2021-08-09   
2  Inactive (no longer requesting services)  2021-06-27   
3     Inactive (inappropriate for services)  2021-06-11   
4                                    Active  2020-12-29   

                       Time Enrolled in START  \
0  Business Hours (Monday - Friday 8am - 5pm)   
1  Business Hours (Monday - Friday 8am - 5pm)   
2  Business Hours (Monday - Friday 8am - 5pm)   
3  Business Hours (Monday - Friday 8am - 5pm)   
4  Business Hours (Monday - Friday 8am - 5pm)   

        Source of referral to START Suitability of enrollment in START  ...  \
0  Case Manager/Service Coordinator                        Appropriate  ...   
1  Case Manager/Service Coordinator                        Appropriate  ...   
2  Case Manager/Service Coordinator                        Appropriate  ...   
3  Case Manager/Service Coordinator                        Appropriate  ...   
4  Case Manager/Service Coordinator                        Appropriate  ...   

  State sponsored ICF/DD stay in past 1 - 5 years (at enrollment to START)  \
0                                                 No                         
1                                                 No                         
2                                                 No                         
3                                                 No                         
4                                                 No                         

  Jailed in the past year (at enrollment to START)  \
0                                               No   
1                                               No   
2                                               No   
3                                               No   
4                                               No   

  Jailed in the past 1 - 5 years (at enrollment to START)  \
0                                                 No        
1                                                 No        
2                                                 No        
3                                                 No        
4                                                 No        

  Employment Status at Enrollment Grade In School      Current IEP  \
0                         Student         Grade 7  Yes: 2020-11-16   
1                         Student        Grade 11               No   
2                         Student         Grade 8  Yes: 2020-01-01   
3                    Not employed             NaN              NaN   
4                             NaN             NaN              NaN   

  Current 504 Plan                                School Setting  \
0               No                Special school-out of district   
1               No                            Other: BOCES 4:1:2   
2               No                    Special school-in district   
3              NaN                                           NaN   
4              NaN  Other: COE Program Held at Cabrillo College    

     Custody Status Child Lives With  
0  Parental custody          Parents  
1  Parental custody          Parents  
2  Parental custody          Parents  
3               NaN              NaN  
4               N

Index(['Local ID', 'Region', 'County', 'Start Coordinator',
       'Date Enrolled in START', 'Status', 'Status Date',
       'Time Enrolled in START', 'Source of referral to START',
       'Suitability of enrollment in START',
       'Individual/Caregiver reliable access to technology',
       'Presenting problems at time of enrollment', 'Services at Enrollment',
       'Date of birth', 'Gender', 'Race', 'Ethnicity',
       'Level of Intellectual Disability',
       'Current living situation at enrollment to START',
       'More than one living situation over the last 5 years?',
       'More than one living situation over the last year?',
       'Psychiatric hospitalization in past year (at enrollment to START)',
       'Psychiatric hospitalizations in past 1 - 5 years (at enrollment to START)',
       'Emergency department visits in past year (at enrollment to START)',
       'Emergency department visits in past 1-5 years (at enrollment to START)',
       'Law enforcement encounters i

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4986 entries, 0 to 4985
Data columns (total 48 columns):
 #   Column                                                                                  Non-Null Count  Dtype         
---  ------                                                                                  --------------  -----         
 0   Local ID                                                                                4986 non-null   object        
 1   Region                                                                                  4986 non-null   object        
 2   County                                                                                  4986 non-null   object        
 3   Start Coordinator                                                                       4947 non-null   object        
 4   Date Enrolled in START                                                                  4986 non-null   datetime64[ns]
 5   Status              

In [5]:
# Explore the presenting problems dataframe
presenting_problems.head()
presenting_problems.columns
presenting_problems.info()

Client Local Id START Coordinator                         Region  \
0               1     Ashley Gruver  Iowa : County Social Services   
1               1     Ashley Gruver  Iowa : County Social Services   
2               1     Ashley Gruver  Iowa : County Social Services   
3         000083W          Amy Huff          North Carolina : West   
4         000083W          Amy Huff          North Carolina : West   

                   Presenting Problems at Enrollment  
0  Aggression (physical, verbal, property destruc...  
1                             Mental health symptoms  
2                                    Self-injurious   
3  Aggression (physical, verbal, property destruc...  
4  Decrease in ability to participate in daily fu...

Index(['Client Local Id', 'START Coordinator', 'Region',
       'Presenting Problems at Enrollment'],
      dtype='object')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31014 entries, 0 to 31013
Data columns (total 4 columns):
 #   Column                             Non-Null Count  Dtype 
---  ------                             --------------  ----- 
 0   Client Local Id                    31014 non-null  object
 1   START Coordinator                  29435 non-null  object
 2   Region                             31014 non-null  object
 3   Presenting Problems at Enrollment  31014 non-null  object
dtypes: object(4)
memory usage: 969.3+ KB


Let's skim `Presenting Problems at Enrollment` to see some of the presented problems

In [6]:
random.choices(presenting_problems['Presenting Problems at Enrollment'].unique(), k = 20)

['Other: Fire starting',
 'Other: Identify appropriate therapist',
 'Other: Training Needed for Providers',
 'Other: medication difficulty',
 'Other: minimal social support, hx of trauma',
 'Other: chronic MHAs',
 'Other: Toilet training regression',
 'Other: loss of school transportation',
 'Other: challenges with taking things from others',
 'Other: Does not want to leave the home',
 'Other: transition from RI to NH',
 'Other: Recent hospitalization',
 'Other: Agoraphobia',
 'Other: urinating on himself',
 'Other: currently in nursing home',
 'Other: Transition out of high school and into independent living in 2021',
 'Other: Day Program needs assistance',
 'Other: med review',
 'Other: Alcohol abuse',
 'Other: Recently lose his job.']

<font color = "green"> As we can see above the presented problems are all over the place. In the next section, we will focus on cleaning this column to narrow down the presented problems.  </font>

# Data cleaning/preprocessing

Let's first rename `Presenting Problems at Enrollment` column to `presented_problems` for more readability.

In [7]:
# rename the column
presenting_problems.rename(columns={'Presenting Problems at Enrollment': 'presented_problems'}, inplace=True)

# confirm that the column was renamed successfully
presenting_problems.columns

Index(['Client Local Id', 'START Coordinator', 'Region', 'presented_problems'], dtype='object')

In [8]:
presenting_problems['presented_problems'].value_counts()

Aggression (physical, verbal, property destruction, threats)                                                                                        7427
Family needs assistance                                                                                                                             4489
Mental health symptoms                                                                                                                              4356
Self-injurious                                                                                                                                      2640
Diagnosis and treatment plan assistance                                                                                                             2246
                                                                                                                                                    ... 
Other: Social Skills                                                              

Now, let's turn all the presenting problems into lowercase to see add consistency.

In [9]:
presenting_problems['presented_problems'] = presenting_problems['presented_problems'].str.lower()
presenting_problems.sample(2)

Client Local Id START Coordinator                          Region  \
14241          327915   Katrina Fullmer             New York : Region 1   
29297         8255457     Mary Akhuetie  California : CA START East Bay   

             presented_problems  
14241  transition from hospital  
29297   family needs assistance

In [10]:
# subset to rows that do not contain 'other:' for presented problems
presenting_problems_no_other = presenting_problems[~presenting_problems["presented_problems"].str.contains('other:')].copy()
presenting_problems_no_other.presented_problems.unique()

print(f"The number of rows in the presenting_problems dataframe before dropping 'other' is: {presenting_problems.shape[0]}")
print(f"The number of rows in the presenting_problems dataframe after dropping 'other' is: {presenting_problems_no_other.shape[0]}")
print(f"The number of unique presenting problems before dropping 'other' is: {presenting_problems.presented_problems.nunique()}")
print(f"The number of unique presenting problems after dropping 'other' is: {presenting_problems_no_other.presented_problems.nunique()}")

array(['aggression (physical, verbal, property destruction, threats)',
       'mental health symptoms', 'self-injurious ',
       'decrease in ability to participate in daily functions',
       'leaving unexpectedly', 'diagnosis and treatment plan assistance',
       'family needs assistance', 'at risk of losing placement',
       'sexualized behavior', 'suicidal ideation',
       'transition from hospital', 'suicidal action'], dtype=object)

The number of rows in the presenting_problems dataframe before dropping 'other' is: 31014
The number of rows in the presenting_problems dataframe after dropping 'other' is: 30011
The number of unique presenting problems before dropping 'other' is: 704
The number of unique presenting problems after dropping 'other' is: 12


Let's see how many rows we are losing by removing presented problems that start with `other:`

In [11]:
print(f"Retention Rate After Dropping 'Other': {round(100*presenting_problems_no_other.shape[0]/presenting_problems.shape[0], 2)}%")

Retention Rate After Dropping 'Other': 96.77%


A retention rate of 96.77% is really good, so we will stick with the 12 unique presented problems. Now, we will write a function to parse the `Region` variable to `state` and `region`

In [12]:
def parse_region(row):
    """
    Parse a string in the format "<state>: <region>" and return a tuple
    containing the state and region.
    
    Parameters:
    row (str): A string in the format "<state> : <region>".
    
    Returns:
    A tuple containing the parsed state and region.
    """
    state, region = row.lower().split(':')
    return state.strip(), region.strip()

# apply the function on the demographics, presenting_problems_no_other, and emergency_crisis_services dataframes
presenting_problems_no_other.loc[:, ['state', 'region_new']] = presenting_problems_no_other['Region'].apply(parse_region).tolist()
demographics.loc[:, ['state', 'region_new']] = demographics['Region'].apply(parse_region).tolist()

# sample from the presenting_problems_no_other dataframe
presenting_problems_no_other.sample(3)

Client Local Id       START Coordinator  \
29875         874735C  Shelle Oosting-Goodwin   
26427         6580564          Alexandra Roth   
7738           164677       Michael Korzelius   

                                  Region  \
29875           North Carolina : Central   
26427  California : CA START San Andreas   
7738                 New York : Region 1   

                            presented_problems           state  \
29875                  family needs assistance  north carolina   
26427                   mental health symptoms      california   
7738   diagnosis and treatment plan assistance        new york   

                 region_new  
29875               central  
26427  ca start san andreas  
7738               region 1

In [13]:
# check all the unique states in the presenting_problems_no_other dataframe
presenting_problems_no_other.state.unique()

array(['iowa', 'north carolina', 'texas', 'new york', 'california',
       'colorado', 'new hampshire', 'maryland'], dtype=object)

Next, we will merge the `demographic` and `presenting_problems_no_other` dataframes using the `Client ID`.

In [14]:
demographics_problems_merged = presenting_problems_no_other.merge(demographics, left_on='Client Local Id', right_on='Local ID',
                                                                  how='inner', suffixes=('_presenting_problems', '_demographics'), 
                                                                  indicator = True).drop_duplicates()

print(f'the shape of the presenting_problems_no_other dataframe is {presenting_problems_no_other.shape}')
print(f'the shape of the demographics dataframe is {demographics.shape}')
print(f'the shape of the demographics_problems_merged dataframe is {demographics_problems_merged.shape}')
print(f'the retention rate after merging the demographics and presenting_problems_no_other dataframes is {round(100*demographics_problems_merged.shape[0]/presenting_problems_no_other.shape[0], 2)}%')

the shape of the presenting_problems_no_other dataframe is (30011, 6)
the shape of the demographics dataframe is (4986, 50)
the shape of the demographics_problems_merged dataframe is (17808, 57)
the retention rate after merging the demographics and presenting_problems_no_other dataframes is 59.34%


We can see that the rows for the `presenting_problems_no_other` is reduced from 30,011 to 17,808. We have lost about 40.66% of our data during the merge process. This is due to the lack of match in the clients' unique ID in the demographic dataset.

In [15]:
demographics_problems_merged.sample(3)
demographics_problems_merged.shape

Client Local Id START Coordinator      Region_presenting_problems  \
16135         7615473   Tracee Williams  California : CA START Westside   
13379          452024      Ebony Harris             New York : Region 1   
11906          422324       Sean Latasa             New York : Region 3   

         presented_problems state_presenting_problems  \
16135  leaving unexpectedly                california   
13379       self-injurious                   new york   
11906   sexualized behavior                  new york   

      region_new_presenting_problems Local ID             Region_demographics  \
16135              ca start westside  7615473  California : CA START Westside   
13379                       region 1   452024             New York : Region 1   
11906                       region 3   422324             New York : Region 3   

            County Start Coordinator  ... Employment Status at Enrollment  \
16135  Los Angeles   Tracee Williams  ...                             NaN   
13379       Monroe      Ebony Harris  ...                         Student   
11906  Westchester       Sean Latasa  ...                         Student   

      Grade In School      Current IEP Current 504 Plan  \
16135             NaN              NaN              NaN   
13379             NaN              NaN              NaN   
11906         Grade 8  Yes: 2019-09-01              NaN   

                       School Setting    Custody Status Child Lives With  \
16135                             NaN  Parental custody          Parents   
13379                             NaN  Parental custody          Parents   
11906  Special school-out of district               NaN              NaN   

      state_demographics region_new_demographics _merge  
16135         california       ca start westside   both  
13379           new york                region 1   both  
11906           new york                region 3   both  

[3 rows x 57 columns]

(17808, 57)

In [16]:
# convert to datetime format and save in enrollment_date
demographics_problems_merged['enrollment_date'] = pd.to_datetime(demographics_problems_merged['Date Enrolled in START'])
print(f"Data type for the date variable: {demographics_problems_merged['enrollment_date'].dtype}")

# round down to the first of the month and year
demographics_problems_merged['enrollment_ym_derived'] = demographics_problems_merged.enrollment_date.dt.to_period('M').dt.to_timestamp(how='start')

# create a new column for the year
demographics_problems_merged['enrollment_year'] = pd.DatetimeIndex(demographics_problems_merged['enrollment_date']).year

demographics_problems_merged[['enrollment_date', 'enrollment_ym_derived', 'enrollment_year']].sample(3)

Data type for the date variable: datetime64[ns]


enrollment_date enrollment_ym_derived  enrollment_year
280        2017-05-30            2017-05-01             2017
14088      2020-11-18            2020-11-01             2020
4155       2017-08-29            2017-08-01             2017

In [17]:
# renaming the disability level column
demographics_problems_merged.rename(columns={'Level of Intellectual Disability': 'disability_level'}, inplace=True)

# remove the rows where 'disability_level' is 'None noted'
demographics_problems_merged = demographics_problems_merged[demographics_problems_merged["disability_level"] != 'None noted']
demographics_problems_merged.disability_level.value_counts()

Mild                   7883
Moderate               4927
Severe                 1250
Normal intelligence    1211
Borderline              982
Profound                220
Name: disability_level, dtype: int64

In [18]:
# make the disability level column an ordered factor variable
disability_levels_ordered = ['Normal intelligence', 'Borderline', 'Mild', 'Moderate', 'Severe', 'Profound']

# convert disability_level to ordered categorical variable
demographics_problems_merged['disability_level_ordered'] = pd.Categorical(demographics_problems_merged['disability_level'], 
                                                                          categories = disability_levels_ordered, ordered = True)

# check the data type of the disability_level_ordered column
demographics_problems_merged['disability_level_ordered'].dtype

CategoricalDtype(categories=['Normal intelligence', 'Borderline', 'Mild', 'Moderate',
                  'Severe', 'Profound'],
, ordered=True)

In [19]:
# remove the rows where 'Gender' starts with 'other'
demographics_problems_merged = demographics_problems_merged[~demographics_problems_merged["Gender"].str.contains('Other:')]
demographics_problems_merged.Gender.value_counts()

Male      11041
Female     5366
Name: Gender, dtype: int64

In [20]:
print(f'There are {len(demographics_problems_merged.columns)} columns in the demographics_problems_merged dataframe.')

There are 61 columns in the demographics_problems_merged dataframe.


In [21]:
demographics_problems_merged.columns

Index(['Client Local Id', 'START Coordinator', 'Region_presenting_problems',
       'presented_problems', 'state_presenting_problems',
       'region_new_presenting_problems', 'Local ID', 'Region_demographics',
       'County', 'Start Coordinator', 'Date Enrolled in START', 'Status',
       'Status Date', 'Time Enrolled in START', 'Source of referral to START',
       'Suitability of enrollment in START',
       'Individual/Caregiver reliable access to technology',
       'Presenting problems at time of enrollment', 'Services at Enrollment',
       'Date of birth', 'Gender', 'Race', 'Ethnicity', 'disability_level',
       'Current living situation at enrollment to START',
       'More than one living situation over the last 5 years?',
       'More than one living situation over the last year?',
       'Psychiatric hospitalization in past year (at enrollment to START)',
       'Psychiatric hospitalizations in past 1 - 5 years (at enrollment to START)',
       'Emergency department visit

Since the dataset has an overwhelming number of columns, we will subset the data to only the columns we will need moving forward.

In [22]:
# subset the demographics_problems_merged dataframe to the columns we need
clean_merged = demographics_problems_merged[['Local ID', 'state_demographics', 'enrollment_date', 
                                             'enrollment_ym_derived', 'enrollment_year', 'disability_level',
                                             'disability_level_ordered', 'presented_problems',
                                             'Gender', 'Ethnicity', 'County']]

clean_merged.head(3)
print(f'The dataset changed from demographics_problems_merged with {demographics_problems_merged.shape[1]} columns to clean_merged with {clean_merged.shape[1]} columns.')

Local ID state_demographics enrollment_date enrollment_ym_derived  \
0  000083W     north carolina      2018-06-27            2018-06-01   
1  000083W     north carolina      2018-06-27            2018-06-01   
2  000083W     north carolina      2018-06-27            2018-06-01   

   enrollment_year disability_level disability_level_ordered  \
0             2018             Mild                     Mild   
1             2018             Mild                     Mild   
2             2018             Mild                     Mild   

                                  presented_problems  Gender  \
0  aggression (physical, verbal, property destruc...  Female   
1  decrease in ability to participate in daily fu...  Female   
2                               leaving unexpectedly  Female   

                Ethnicity County  
0  Not of Hispanic origin  Swain  
1  Not of Hispanic origin  Swain  
2  Not of Hispanic origin  Swain

The dataset changed from demographics_problems_merged with 61 columns to clean_merged with 11 columns.


In [23]:
# save the cleaned_merged dataframe to a pkl file
clean_merged.to_pickle('../data/clean_presented_problems.pkl')